GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -492,6 +522,75 @@ private bool TryRecursivelyDeleteInstallerDirectory(out string error)
             return true;
         }
 
+        private IList<IPackageSearchMetadata> QueryFeedWithRetry()
+        {
+            int retryCount = 0;
+            IList<IPackageSearchMetadata> queryResults = null;
+
+            do
+            {
+                try
+                {
+                    queryResults = this.nuGetFeed.QueryFeedAsync(this.nuGetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+                    return queryResults;
+                }
+                catch (Exception ex) when (ex is FatalProtocolException &&
+                                           ex.InnerException is System.Net.Http.HttpRequestException &&
+                                           (ex.InnerException.Message.Contains("401") || ex.InnerException.Message.Contains("403")) &&
+                                           retryCount == 0)
+                {
+                    // If we fail to query the feed due to an authorization error, then it is possible we have stale
+                    // credentials, or credentials without the correct scope. Re-aquire fresh credentials and try again.
+                    EventMetadata data = CreateEventMetadata(ex);
+                    this.tracer.RelatedWarning(data, "Failed to query feed due to unauthorized error. Re-acquiring new credentials and trying again.");
+
+                    retryCount++;
+
+                    if (!this.TryRefreshCredentials(out string error))
+                    {
+                        // If we were unable to re-acquire credentials, throw a new exception indicating that we tried to handle this, but were unable to.
+                        throw new Exception($"Failed to query the feed for upgrade packages due to: {ex.Message}, and was not able to re-acquire new credentials due to: {error}", ex);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    EventMetadata data = CreateEventMetadata(ex);
+                    string message = $"Error encountered when querying NuGet feed. Number of attempts: {retryCount + 1}";
+                    this.tracer.RelatedWarning(data, message);
+                    throw new Exception($"Failed to query the NuGet package feed due to error: {ex.Message}", ex);
+                }

```

Comment:
Do we even need this catch if we are rethrowing and it will only happen on a retry.  The call stack should give us the information we need when it is caught higher up in the call stack right?

Reply:
The excepton is caught here because the underlying exception can be quite ambiguous and unhelpful to a user. The idea was put some context around the unhelpful error (e.g. "Cannot Query the NuGet Feed"). 

In practice, I think what I have still falls short, but I thought I would tackle that separately.